modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/ReadWriteTransactionImpl.java

```diff
@@ -150,26 +150,39 @@ private void checkEnlistPossibility() {
     @Override
     public CompletableFuture<Void> commitAsync() {
         return TransactionsExceptionMapperUtil.convertToPublicFuture(
-                finish(true, null, false),
+                finish(true, null, false, false),
                 TX_COMMIT_ERR
         );
     }
 
     @Override
     public CompletableFuture<Void> rollbackAsync() {
         return TransactionsExceptionMapperUtil.convertToPublicFuture(
-                finish(false, null, false),
+                finish(false, null, false, false),
                 TX_ROLLBACK_ERR
         );
     }
 
     @Override
-    public CompletableFuture<Void> finish(boolean commit, @Nullable HybridTimestamp executionTimestamp, boolean full) {
+    public CompletableFuture<Void> rollbackTimeoutExceededAsync() {
+        return TransactionsExceptionMapperUtil.convertToPublicFuture(
+                finish(false, null, false, true)
+                        .thenAccept(unused -> timeoutExceeded = true),
+                TX_ROLLBACK_ERR
+        ).thenAccept(unused -> this.timeoutExceeded = true);
```

Comment:
Why does this assignment happen after the finish gets replicated? Also, can there be a race with finish()?

Reply:
good catch!

Reply:
I think this flag should be set somethere inside finis(..., true)
See how this `org.apache.ignite.internal.tx.impl.ReadWriteTransactionImpl#killed` 
is implemented

Reply:
Than makes sense, thank you.